keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,553 @@
+feedback or metrics.
+- Providing a way for kube-proxy to consume subsets of EndpointSlices (this is
+a separate KEP).
+- Providing a way to require Node local traffic ([this is a separate KEP](https://github.com/kubernetes/enhancements/pull/1944)).
+- Multi-cluster topology aware routing (this same pattern may be useful there
+though).
+- Region based topology aware routing (this may come later).
+- Ensuring that endpoints are distributed evenly across zones.
+
+## Proposal
+
+### Terminology
+
+- **EndpointSlice Group**: One or multiple EndpointSlices that will be consumed
+by the same set of zones.
+- **Endpoints**: In this KEP, “endpoints” are used to describe individual Pod
+IPs. These are not references to Kubernetes Endpoints resources. We also omit
+the unready pods that number of endpoints = number of pods.
+
+### Configuration
+
+This proposal builds off of the [EndpointSlice API](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/).
+
+We propose a new Service annotation: `endpointslice.kubernetes.io/same-zone`
+with three options:
+- `Balanced`:This is the default, it does not need to be specified which also
+provides backward compatibility for users. It represents the existing approach
+that is already used, no special mechanism will be added. Traffic will be
+evenly distributed among endpoints behind a Service.
+- `Require`: EndpointSlice controller will conduct a simple but aggressive
+routing mechanism. It will group endpoints in a zone to an EndpointSlice Group
+that only receives traffic from that zone. `Require` requires users to ensure
+that a sufficient number of endpoints for a Service exist in each zone.
+Services using this approach will only be reachable from zones that contain
+endpoints for the Service.
+- `Prefer`: EndpointSlice controller will conduct a more conservative routing
+mechanism. It will keep traffic in zone as long as each zone has a sufficient
+number of endpoints. If a zone does not have a sufficient number of endpoints,
+traffic will be routed to other zones.
+
+This same pattern could be used for topology aware routing at node and region
+level. We could add new annotations `endpointslice.kubernetes.io/same-node` and
+`endpointslice.kubernetes.io/same-region` for node and region respectively with
+the same values `Require` and `Prefer`. With all these three levels combined
+together, their individual value could be influenced by values of others, i.e.
+`Require` node would also `Require` zone and region.
+
+
+### Notes/Constraints/Caveats
+
+In the future, we may introduce another parameter for the `Prefer` approach
+that would allow users to specify a starting threshold. Starting threshold
+defines the starting point for `Prefer` (more detailed description below). We
+currently set the starting threshold to 3x the number of zones based on results
+from our test dataset for safer distribution results.
+
+### Risks and Mitigations
```

Comment:
it also will add an overhead to the endpointslice controller that may lag to update other endpoint slices, is this a scalability risk?

Reply:
Yeah that's a good point. The added complexity could result in slower updates and end up limiting scalability (hopefully not though).

Reply:
Added the concern to this section.